### ISSUE TITLE
warning: pointless comparison of unsigned integer with zero

### ISSUE THREAD

**Issue by psychocoderHPC** [NVIDIA/cub#343]
As soon as `cub` (shipped with CUDA toolkit) headers are included in my mini reproducer nvcc throw the warning
`warning: pointless comparison of unsigned integer with zero`.
This example is simplified, in the original code `foo` is calculated at compile time based on user-provided values.
The compiler is right that `i < foo` is always `false`, and a useless check, but this expression should not be checked if `foo != 0` is false.
I am wondering why including `cub` is enabling this warning.

```C++
#include <iostream>
#include <cstdint>
#ifdef WITH_CUB
#  include <cub/cub.cuh>
#endif

__global__ void kernel(uint32_t i){
   constexpr uint32_t foo = 0u;
   if(foo != 0u && i < foo)
         printf("x\n");
}

int main(){
        return 0;
}
```

`nvcc  -x cu main.cpp` is not throwing a warning because of `foo != 0` the comparison `i < foo` should not be evaluated.
`nvcc  -x cu main.cpp -DWITH_CUB` is throwing:
```
main.cpp(9): warning: pointless comparison of unsigned integer with zero

main.cpp(9): warning: pointless comparison of unsigned integer with zero
```

Tested with : nvcc 11.X (X=1,2,3)

**Comment by psychocoderHPC** [comment]
The mini example is derived from [this compile-time loop in PIConGPU]( 
https://github.com/ComputationalRadiationPhysics/picongpu/blob/c71dbe451a12e1045685dec3a88eb6d1339ff405/include/pmacc/lockstep/ForEach.hpp#L140)

**Comment by senior-zero** [comment]
I can confirm this behaviour. CUB uses `<thrust/system/cuda/detail/core/triple_chevron_launch.h>` which includes `thrust/functional.h` -> `thrust/detail/functional/operators.h`  -> `thrust/detail/functional/operators/logical_operators.h` which contains:

```cuda
template<typename T1, typename T2>
__host__ __device__
actor<
  composite<
    transparent_binary_operator<thrust::logical_and<>>,
    actor<T1>,
    typename as_actor<T2>::type
  >
>
operator&&(const actor<T1> &_1, const T2 &_2)
{
  return compose(transparent_binary_operator<thrust::logical_and<>>(),
                 make_actor(_1),
                 make_actor(_2));
} // end operator&&()
```

So, your example can be actually simplified to:
```cuda
#include <iostream>
#include <cstdint>

template<typename T1, typename T2>
int operator&&(const T1 &_1, const T2 &_2)
{
  return {};
}

__global__ void kernel(uint32_t i){
  constexpr uint32_t foo = 0u;
  if(foo != 0u && i < foo)
    printf("x\n");
}

int main(){
  return 0;
}
```

I don't quite understand why the warning is triggered in this case, so we'll discuss this issue later. For now, I can suggest using `#include <cub/{block, warp}/*>` directly instead of `#include <cub/cub.cuh>` if it's possible in your case.

**Comment by psychocoderHPC** [comment]
Thanks, for me the workaround of including `#include <cub/block/block_radix_sort.cuh>` to use radix sort within a kernel works.

**Comment by allisonvacanti** [comment]
I suspect that the `operator&&` is forcing the arguments to be evaluated, breaking the "short circuit" behavior we'd normally expect here. 

We can likely fix this by making the `operator&&` a hidden friend of `actor`. That way it won't leak into the global namespace.

**Comment by allisonvacanti** [comment]
Internal NVBug 3341536 may be related, which describes a compiler bug with the same behavior.

**Comment by allisonvacanti** [comment]
This was a bug in nvcc and was fixed in CTK 11.5.
